Added sys:cron:schedule command

hostep opened the pull request:
Magerun pull-request check-list:

- [x] Pull request against develop branch (if not, just close and create a new one against it)
- [ ] README.md reflects changes (if any)

Sometimes we want to run a cronjob, but not keep our ssh session or terminal open while running `sys:cron:run`. This PR adds a `sys:cron:schedule` command which schedules a certain job in the cron schedule to be executed the next time the cron is being run.
Was inspired by the excellent [Aoe_Scheduler module](https://github.com/AOEpeople/Aoe_Scheduler) for M1.

Possible discussion points:
1) The name `schedule` might be confusing, people might expect they can provide a timestamp on which time they want to schedule a job. Is this wanted? We could also name the command `schedule-now` or something like that?

2) I copied a lot of code from the `RunCommand` class. This should definitely be refactored away. Should we put shared code in the `AbstractCronCommand` class, or someplace else?

3) There is something wrong with the timezone of the times in the `RunCommand` class. I haven't fixed this yet over there, but this new class should work correctly. I took a look at how [M2 itself does this](https://github.com/magento/magento2/blob/056ad90b7ee3177d1665d39c86191ac156bff697/app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php#L443-L478) and based my code on that.

If there are any remarks, let me know!

ktomk commented on src/N98/Magento/Command/System/Cron/AbstractCronCommand.php:
```diff
@@ -50,6 +57,7 @@ public function inject(
         $this->_cronConfig = $cronConfig;
         $this->_scopeConfig = $scopeConfig;
         $this->_cronScheduleCollection = $cronScheduleCollection;
+        $this->_timezone = $timezone;
```
Please convert ->_... properties to ones w/o "_" at the beginning. I know it's now all your code, so please at least do it for new introduced properties even if code-style changes that way (it changes to the correct way).

ktomk commented on src/N98/Magento/Command/System/Cron/ScheduleCommand.php:
```diff
@@ -0,0 +1,107 @@
+        $schedule
+            ->setJobCode($jobCode)
+            ->setStatus(Schedule::STATUS_PENDING)
+            ->setCreatedAt(strftime('%Y-%m-%d %H:%M:%S', $this->_timezone->scopeTimeStamp()))
```
created at with seconds, good.

ktomk commented on src/N98/Magento/Command/System/Cron/ScheduleCommand.php:
```diff
@@ -0,0 +1,107 @@
+            ->setJobCode($jobCode)
+            ->setStatus(Schedule::STATUS_PENDING)
+            ->setCreatedAt(strftime('%Y-%m-%d %H:%M:%S', $this->_timezone->scopeTimeStamp()))
+            ->setScheduledAt($scheduledAtTime)
```
I think this needs to be a formatted timestamp as well, correct me if not.

ktomk commented on src/N98/Magento/Command/System/Cron/ScheduleCommand.php:
```diff
@@ -0,0 +1,107 @@
+     */
+    protected function askJobCode(InputInterface $input, OutputInterface $output, $jobs)
+    {
+        foreach ($jobs as $key => $job) {
```
Sadly we don't offer anything less cumbersome here. This might be something worth in a base-class if shared with the other command as asking is command related (less cron sheduling related, at least for most of this code).

ktomk commented:
> Possible discussion points:
> 1) The name `schedule` might be confusing, people might expect they can provide a timestamp on which time they want to schedule a job. Is this wanted? We could also name the command `schedule-now` or something like that?
AFAIK we can't do that as in the end Magento decides when to execute it and it depends on how many there are. There is a timestamp so it can be set, but remeber it must not be executed at that time. It's something line "not before" *that* timestamp. Maybe defer this for a moment.

For the wording, please also take a look with Magerun 1 where - IIRC - a cron schedule command exists. In any case this should be backwards compatible. If this does not match, let's defer the details. But if you can, take a look.

> 2) I copied a lot of code from the `RunCommand` class. This should definitely be refactored away. Should we put shared code in the `AbstractCronCommand` class, or someplace else?

If you can refactor it into a model class, I'd say it's preferable over an abstract base class as we should favor aggregation over inheritance. You can place such a model class next to those command classes (in the same directory). This might take more time than pushing members down to a new abstract-base-class. You're open to do it the way you prefer (or what currently works best for you), so no hard rules here. Historically, abstract classes were preferred, for me personally that's not my style but just find your way that is practically for you right now.

> 3) There is something wrong with the timezone of the times in the `RunCommand` class. I haven't fixed this yet over there, but this new class should work correctly. I took a look at how [M2 itself does this](https://github.com/magento/magento2/blob/056ad90b7ee3177d1665d39c86191ac156bff697/app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php#L443-L478) and based my code on that.
I have no decision on this yet. I've seen the last ~3-5 days that Magento uses `strftime` which is locale dependent so I can imagine (but have not checked as only just seen) that locale might incorporate the timezone for the timestamp (sorry no definite check on it, references welcome). If that (locale and timezone dependent timestamp) is the case it would violate the best practice to use UTC based timestamps (not timezone dependent ones). So bringing inconsistencies in the mix (locale is not thread-safe) this might require the work-around you linked.

From this quick glance over the last days (was looking in there because of some Magerun 1 cron command), I'd say the use of strftime within Magento core definitely is a smell and I wondered why Magento core makes use of it for timestamps stored into the database system. Really.

For your question: I would prefer to not duplicate any work-arounds (if there are any) but instead do as documented: `$time` is formatted by `strftime()` and so be it. If then by usage of the command some inconsistency turns out, we can take a closer look into it. But I'd say it should just work this way, nothing complicated necessary (e.g. `self::SECONDS_IN_MINUTE` is bogus for correct time interval calculations when interested in iterating minutes correctly - imagine extra seconds for some days of the year- however on duplicates it's just continued two times on the same timestamp). Just do a good created at (that one is with seconds) and for the shedule w/o seconds. I've left a comment in review.

Generally it would be great if you could increase code-coverage. Code looks generally good, I've left some quick comments. If you want to already extract functionality, you can do so if you like but it's not a necessity. IMHO code coverage is more important so that it's possible to refactor more easily later on.

Thanks a lot for you PR. If you have any feedback on my feedback just shoot!

hostep commented:
Hi @tkn98 

Sorry for the delay, but I just found time to update this PR and I think I addressed most of your remarks. If you have new remarks, please let me know.

> If you can refactor it into a model class, I'd say it's preferable over an abstract base class as we should favor aggregation over inheritance.

I chose to move the `askJobCode` method to the `AbstractCronCommand` class, because it would take me to long to refactor it decently using composition over inheritance.

> For the wording, please also take a look with Magerun 1 where - IIRC - a cron schedule command exists. In any case this should be backwards compatible. If this does not match, let's defer the details. But if you can, take a look.

As far as I can see, there is no such command in Magerun 1, so the naming is still open for discussion :)

ktomk commented:
Thanks, and sorry for the delay, too :) This is now in the develop version.

And for the name: I don't know where I took that from, so, no discussion at all ^ ^
